### ISSUE TITLE
Avoid re-fragmenting memtable range tombstones on each read

### ISSUE THREAD

**Issue by abhimadan** [facebook/rocksdb#4808]
Each time `MemTable::NewRangeTombstoneIterator` is called, the tombstones in the dedicated range deletion memtable are fragmented. Range deletions are probably issued as frequently as other forms of writes, so it should be possible to do some more work to avoid this in the average case. The fragmented tombstones can be cached, though I think doing more work on the write path is preferable, since we currently do the bare minimum of just writing a single entry to the memtable.

**Comment by jwasinger** [comment]
Do I understand this correctly?  Fragmented tombstones should computed during write and cached.  Reads would reference the cache instead of recomputing. 

> more work on the write path is preferable

@abhimadan Can you elaborate?  Is this different than what I have proposed above?

**Comment by abhimadan** [comment]
What I meant by caching the fragmented tombstones was that, during the first read after the range deletion, the fragmented memtable tombstones would be stored and used for subsequent reads until the next range deletion invalidates the cache. So basically the "caching" happens on the read path.

On the write path, I was thinking of using a data structure that would fragment a tombstone appropriately after it is written. So instead of doing the fragmentation in a separate pass and caching the result, we could go even further and the fragment the tombstones in-place, which I expect to be cheaper, as it wouldn't require re-creating a new data structure every range deletion. The tricky part is that this would modify entries already in the memtable, which the skiplist memtable forbids. @ajkr had an idea to make a special range deletion memtable that would give each range tombstone fragment entry two sequence numbers, to allow tombstones inserted into the memtable to either represent new tombstones or to represent fragments of existing tombstones (I think we may have a more detailed description of this somewhere if you're interested).

TL;DR I was proposing either caching would be on the read path or an in-place fragmented data structure on the write path.

**Comment by jwasinger** [comment]
> (I think we may have a more detailed description of this somewhere if you're interested)

Yes, I would be interested in seeing this.

Read path caching seems like it would be the easiest to implement in the short term.  I have to look more closely at the write path idea.

**Comment by jwasinger** [comment]
@abhimadan @ajkr , I have implemented read "caching" for the fragmented tombstones as a preliminary step towards closing this issue.  Can you link me to the detailed description of the proposal for a special range deletion memtable?  It sounds very interesting.

Thanks,
Jared

**Comment by abhimadan** [comment]
Hi Jared, sorry about the delay. Here's [the doc I was talking about](https://docs.google.com/document/d/1NfjELi8Zz27TUZ3HoxujPRLKe6ifx_Jg9cAEF6WtZNg/edit#heading=h.1w0pjit7yac): (just the "Range Tombstone Data Format" section).

**Comment by jwasinger** [comment]
Thanks @abhimadan :) !
